libraries/node-core-library/src/Async.ts

```diff
@@ -190,3 +190,91 @@ export class Async {
     }
   }
 }
+
+type IterationResolveType = 'push' | 'callback';
+
+/**
+ * An async iterable queue that allows for asynchronous iteration. During iteration,
+ * the queue will wait until the next item is pushed into the queue before yielding.
+ * If instead all queue items are consumed and all callbacks are called, then the
+ * queue will return.
+ *
+ * @public
+ */
+export class AsyncQueue<T> implements AsyncIterable<[T, () => void]> {
+  private _queue: T[];
+  private _onPushPromise: Promise<'push'>;
+  private _onPushResolve: () => void;
+
+  public constructor(iterable?: Iterable<T>) {
+    this._queue = iterable ? Array.from(iterable) : [];
+    const [promise, resolver] = this._getIterationPromise('push');
+    this._onPushPromise = promise;
+    this._onPushResolve = resolver;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public [Symbol.asyncIterator](): AsyncIterator<[T, () => void], any, undefined> {
```

Comment:
You might find it more ergonomic to do:
```ts
public async *[Symbol.asyncIterator](): AsyncIterableIterator<[T, () => void]> {
```
And use `yield` for each intervening value instead of having to worry about the details of the mechanics of `next()` and the `done` flag.